### DISCUSSION TITLE
New LEGO Wireless Protocol v3 tool

### DISCUSSION THREAD

**Discussion by dlech** [pybricks/support#353] [Announcements]
We have just added a new feature to our [pybricksdev](https://github.com/pybricks/pybricksdev) command line tool to interactively send and receive messages to Powered Up hubs that support the [LEGO Wireless Protocol v3](https://lego.github.io/lego-ble-wireless-protocol-docs/).

The tool acts as a REPL that lets you construct arbitrary messages to send to the hub. It also stores the history of commands you entered so you can easily repeat or modify previous commands sent.

https://user-images.githubusercontent.com/963645/120021842-3426b280-bfb1-11eb-9439-a9152ca7e556.mp4

We built this to learn more about how the Powered Up remote control works, but it should work with any Powered Up hub running LEGO firmware.

The message classes and parameter enums (bytecodes) are documented [here](https://docs.pybricks.com/projects/pybricksdev/en/latest/api/ble/lwp3/index.html).

**Comment by dlech** [comment]
Hi @nathankellenicki and @tthiery,

I know Python is not your language of choice, but I think this might be helpful for you just for exploring new devices. And if you are interested and have time, we could collaborate on documenting the LWP3 protocol here based on what we have learned since LEGO doesn't seem to have much time for updating their version of the docs (it might be better than digging through each others source code :wink:).

**Reply by tthiery** [reply]
Thanks for the hint. When I buy my next Powered Up toy without having details, I will have a look at the repl-cli. Your tool looks good ... I have recently thought about a interactive notebook powered C# based repl/controls ... but too much to do irl 😀.

I have started some months ago a wiki-like repository named [sharpbrick/docs](https://github.com/sharpbrick/docs) for the purpose of documenting - library independent - the strange behaviors of the various Powered UP sensors, motors, hubs etc. I had a dozen "investigation" tagged issues in my library so I basically pushed out investigation/knowledge part out of my issues. However, other repositories like that (e.g. [bricklife/LEGO-Mario-Reveng](https://github.com/bricklife/LEGO-Mario-Reveng)) are drying out. I am also in favor of a central place documenting the devices. I think you even contributed, right 😀.

Is you suggestion to (community) re-write the LEGO specification to make it significant more clearer? I would definitely contribute to that. I am not good with investigations, but I am good in doing abstract work and guessing original architecture intent 😀.

Speaking honestly, I am far more concerned about the modern Mindstorms protocol (with whatever future it has). The LWP3 is, like you said via implementations and the original spec, quite well understood. And there the spec-silence of LEGO really kills me. I would really have a private conversation between us and the Product Manager at LEGO about the future of this. I mean our libraries are the ultimate insurance that the these bricks are not bricked bricks.

**Reply by dlech** [reply]
> I have started some months ago a wiki-like repository named [sharpbrick/docs](https://github.com/sharpbrick/docs) for the purpose of documenting

Nice. We've been collecting some of the same info [here](https://github.com/pybricks/technical-info/blob/master/uart-protocol.md) but it isn't quite as extensive. It does seem that these kinds of repos don't get much attention in terms of feedback and contributors.

> Is you suggestion to (community) re-write the LEGO specification to make it significant more clearer?

If people are willing to put in the effort, then I suppose, yes. And since the official LEGO docs are MIT licensed, really we can just copy, tweak and organize instead of a complete re-write. One concern I have though is that if we go though this effort, how do we make it more discoverable than the other GitHub repos that die out or aren't really noticed much at all?

**Reply by tthiery** [reply]
Well, when @nathankellenicki and @corneliusmunz are on board we can prominently link the community spec it in the README docs of the dominant implementations of JS, Python, C++ and .NET .. that will solve the discoverability, awareness and (maybe) contribution part. In the end, each project has 3-10 regular/occasional contributors ... which will contribute. I saw that in my project (and no one ever claimed .NET is the most contribution friendly environment).

Let us get their feedback and then develop a plan to merge the docs.

ps: We can give it a fancy name .. like the LWP3 by the "Guardians of the Electric Bricks". There is so much fun potential in the names around these Powered UP projects. SharpBrick is only 25% about C# but 75% about my bare feet and bricks I run over when my girls do not clean up their room.
